### PR TITLE
User by skills

### PR DESCRIPTION
**Pull request by OlgaEvdokimova**
endpoint get users by skills

### CODE REVIEW

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java [hunk 1]
```diff
@@ -149,4 +153,11 @@ public void updateTimeSlot(UUID userId, List<UserTimeSlotDTO> userTimeSlotDTOLis
         return mapper.map(users);
     }
 
+    @Override
+    public Set<UserDTO> getUsersBySkills(List<Skill> skills) {
+        List<String> skillsToString = skills.stream().map(Enum::toString).collect(Collectors.toList());
```
Comment (Fenydude) [inline src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Please, add log

Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Done, thanks

**Review by ShirshovAlex — DISMISSED** [review]

File: src/main/java/by/exadel/internship/repository/UserRepository.java [hunk 2]
```diff
@@ -40,4 +41,7 @@
     List<User> findAllWithSkillByInternshipId(@Param("id") UUID internshipId, @Param("role") UserRole role);
 
     List<User> findAllByUserRole(UserRole userRole);
+
+    @Query(value = "SELECT * FROM user_detail as u JOIN user_skill as s ON u.u_id = s.us_u_id WHERE CAST (s.us_name AS VARCHAR) IN :skills AND u.u_deleted = false", nativeQuery = true)
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/repository/UserRepository.java]: Why native query?

File: src/main/java/by/exadel/internship/controller/UserController.java [hunk 3]
```diff
@@ -65,6 +69,14 @@ public void restoreUser(@PathVariable UUID userId) {
...
     public void saveUserTime(@PathVariable(name = "userId") UUID userId,
                              @RequestBody List<UserTimeSlotDTO> userTimeSlotDTOList) {
-        userTimeSlotService.saveUserTime(userTimeSlotDTOList,userId);
+        userTimeSlotService.saveUserTime(userTimeSlotDTOList, userId);
     }
+
+    @AdminAccessControl
+    @GetMapping("/skills")
+    @ApiOperation("Get user by skills")
+    public Set<UserDTO> getUserBySkills(@RequestParam("skills") List<Skill> skills) {
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/controller/UserController.java]: What will happen if you pass invalid skill name or valid but with lower case mode for example?

File: src/main/java/by/exadel/internship/repository/UserRepository.java (hunk 2 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/repository/UserRepository.java]: Because user_detail table don't have column skills and we neet to join and find by skill name, and skill is Enum, that is why i decided to create native querry.

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -149,4 +153,14 @@ public void updateTimeSlot(UUID userId, List<UserTimeSlotDTO> userTimeSlotDTOLis
...
+        MDCLog.putClassNameInMDC(SIMPLE_CLASS_NAME);
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Looks like an extra line. There is no need to put class name in MDC.

Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: I dropped it

File: src/main/java/by/exadel/internship/repository/UserRepository.java (hunk 2 above)
Comment (SergeyKoval) [reply src/main/java/by/exadel/internship/repository/UserRepository.java]: ok

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -149,4 +153,14 @@ public void updateTimeSlot(UUID userId, List<UserTimeSlotDTO> userTimeSlotDTOLis
         return mapper.map(users);
     }
 
+    @Override
+    public Set<UserDTO> getUsersBySkills(List<Skill> skills) {
+        List<String> skillsToString = skills.stream().map(Enum::toString).collect(Collectors.toList());
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Minor: each method call from new line, please

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -149,4 +153,14 @@ public void updateTimeSlot(UUID userId, List<UserTimeSlotDTO> userTimeSlotDTOLis
...
+        MDCLog.putClassNameInMDC(SIMPLE_CLASS_NAME);
+        log.info("Try to get set of user by skills : {}", skills);
+        List<User> usersWithSkills = userRepository.getUsersBySkills(skillsToString);
+        Set<User> newList = new HashSet<>(usersWithSkills);
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Why Set? Can you return List?

Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Set because 1 User can have 2 request skills and in the List he we be added 2 times

File: src/main/java/by/exadel/internship/controller/UserController.java (hunk 3 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/UserController.java]: handled it. Check my code, please

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java (hunk 5 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: This line I dropped. But I got you, thanks

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -149,4 +150,19 @@ public void updateTimeSlot(UUID userId, List<UserTimeSlotDTO> userTimeSlotDTOLis
         return mapper.map(users);
     }
 
+    @Override
+    public Set<UserDTO> getUsersBySkills(List<String> skills) {
+        List<String> enumToStringList = Arrays.stream(Skill.values())
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Use better EnumUtils to check

File: src/test/java/by/exadel/internship/userTest/UserTest.java [hunk 8]
```diff
@@ -115,6 +115,19 @@ public void checkListSizeHistoricalUsers() throws Exception {
         assertEquals(usersHistorical.size(), 1);
     }
 
+    @Test
+    public void givenListOfSkills_checkUsersBySkills() throws Exception {
```
Comment (SergeyKoval) [inline src/test/java/by/exadel/internship/userTest/UserTest.java:120]: Test invalid skills as well

File: src/main/java/by/exadel/internship/controller/UserController.java [hunk 9]
```diff
@@ -65,6 +69,14 @@ public void restoreUser(@PathVariable UUID userId) {
...
     public void saveUserTime(@PathVariable(name = "userId") UUID userId,
                              @RequestBody List<UserTimeSlotDTO> userTimeSlotDTOList) {
-        userTimeSlotService.saveUserTime(userTimeSlotDTOList,userId);
+        userTimeSlotService.saveUserTime(userTimeSlotDTOList, userId);
     }
+
+    @AdminAccessControl
+    @GetMapping("/skills")
+    @ApiOperation("Get user by skills")
+    public Set<UserDTO> getUserBySkills(@RequestParam("skills") List<String> skills) {
```
Comment (viktarkolbik) [inline src/main/java/by/exadel/internship/controller/UserController.java:78]: Why did you change from List<Skills> to List<String>?

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (kzaretskaya) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Try to use `distinct` in query.

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java (hunk 7 above)
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: You don't need custom validation of Enum values. If your endpoint receives List<Skills> it will throw BadRequest by default if there will be invalid value. 

 If you follow your logic then you need to check UUID values as  well. Another case for Enums, Why didn't you  check Enums which is sent inside of the request body?

File: src/main/java/by/exadel/internship/controller/UserController.java (hunk 3 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/controller/UserController.java]: we will have 400 status, so we need to write right request , how to write our Skill we have GET endpoint for seeing it

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java (hunk 5 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: I got it back and redone.

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java (hunk 7 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: we talked about this check.

Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: I returned back previous variant. Yes, I got it

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: And I got 2 users the same ones

File: src/test/java/by/exadel/internship/userTest/UserTest.java (hunk 8 above)
Comment (OlgaEvdokimova) [reply src/test/java/by/exadel/internship/userTest/UserTest.java:120]: Done

File: src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java (hunk 6 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/service/impl/UserServiceImpl.java]: Done, thanks

**Review by kzaretskaya — APPROVED** [review]

**Review by ShirshovAlex — APPROVED** [review]
